File: src/main/java/ru/pearx/pmdumper/dumpers/DumperBiomes.java
package ru.pearx.pmdumper.dumpers;

import net.minecraft.entity.EnumCreatureType;
import net.minecraft.util.ResourceLocation;
import net.minecraft.world.biome.Biome;
import net.minecraftforge.fml.common.registry.ForgeRegistries;
import net.minecraftforge.fml.relauncher.Side;
import net.minecraftforge.fml.relauncher.SideOnly;
import ru.pearx.pmdumper.dumpers.base.IDumper;
import ru.pearx.pmdumper.dumpers.base.PMDData;

import java.util.*;

/**
 * Created by mrAppleXZ on 14.06.17 9:36.
 */
@SideOnly(Side.CLIENT)
public class DumperBiomes implements IDumper
{
    @Override
    public String getName()
    {
        return "biomes";
    }

    @Override
    public PMDData getData()
    {
        Map<String, Integer> counts = new HashMap<>();
        List<List<String>> lst = new ArrayList<>();
        List<String> header = new ArrayList<>();
        header.addAll(Arrays.asList("ID", "Name", "Temperature", "Base Height", "Class Name", "Is Snowy", "Can Rain", "Rainfall", "Base Biome", "Filler Block", "Top Block"));
        for(EnumCreatureType t : EnumCreatureType.values())
        {
            header.add("Spawn List (" + t.toString() + ")");
        }
        lst.add(header);


        for(Map.Entry<ResourceLocation, Biome> entr : ForgeRegistries.BIOMES.getEntries())
        {
            ResourceLocation loc = entr.getKey();
            Biome b = entr.getValue();
            List<String> bLst = new ArrayList<>();
            bLst.add(loc.toString());
            bLst.add(b.getBiomeName());
            bLst.add(b.getTemperature() + " (" + b.getTempCategory().toString() + ")");
            bLst.add(Float.toString(b.getBaseHeight()));
            bLst.add(b.getBiomeClass().getName());
            bLst.add(Boolean.toString(b.isSnowyBiome()));
            bLst.add(Boolean.toString(b.canRain()));
            bLst.add(Float.toString(b.getRainfall()));
            bLst.add(b.baseBiomeRegName == null ? "no" : b.baseBiomeRegName);
            bLst.add(b.fillerBlock.toString());
            bLst.add(b.topBlock.toString());
            for(EnumCreatureType t : EnumCreatureType.values())
            {
                bLst.add(spawnableListToString(b.getSpawnableList(t)));
            }
            lst.add(bLst);

            PMDData.plusCounts(counts, loc.getResourceDomain(), 1);
        }

        return new PMDData(lst, counts);
    }

    public String spawnableListToString(List<Biome.SpawnListEntry> lst)
    {
        StringBuilder bld = new StringBuilder();
        for(Biome.SpawnListEntry entr : lst)
        {
            bld.append(entr.toString());
            bld.append(System.lineSeparator());
        }
        if(lst.size() > 0)
            bld.delete(bld.length() - System.lineSeparator().length(), bld.length());
        return bld.toString();
    }
}

File: src/main/kotlin/ru/pearx/pmdumper/dumper/standard/DumperBiomes.kt
@file:JvmMultifileClass
@file:JvmName("StandardDumpers")
package ru.pearx.pmdumper.dumper.standard

import net.minecraft.entity.EnumCreatureType
import net.minecraft.util.ResourceLocation
import net.minecraftforge.fml.common.registry.ForgeRegistries
import ru.pearx.pmdumper.*
import ru.pearx.pmdumper.dumper.dumper
import ru.pearx.pmdumper.utils.client
import ru.pearx.pmdumper.utils.mutableListOfNotNull
import ru.pearx.pmdumper.utils.toHexColorString
import ru.pearx.pmdumper.utils.toPlusMinusString

val DumperBiomes = dumper {
    registryName = ResourceLocation(ID, "biomes")
    header = mutableListOfNotNull("ID", client("Name"), "Default Temperature", "Base Height", "Height Variation", "Class Name", "Is Snowy", "Can Rain", "Rainfall", "Base Biome", "Filler Block", "Top Block", client("Water Color"), "Water Color Multiplier", "Creature Spawning Chance").apply {
        for (type in EnumCreatureType.values())
            add("${type.toString().toLowerCase().capitalize()} Spawn List: Entity*(Min Group-Max Group):Weight")
    }
    amounts {
        for (biome in ForgeRegistries.BIOMES)
            this += biome.registryName
    }
    iterator {
        for (biome in ForgeRegistries.BIOMES) {
            with(ArrayList<String>(header.size)) {
                with(biome) {
                    add(registryName.toString())
                    client { add(biomeName) }
                    add(defaultTemperature.toString())
                    add(baseHeight.toString())
                    add(heightVariation.toString())
                    add(this::class.java.name)
                    add(isSnowyBiome.toPlusMinusString())
                    add(canRain().toPlusMinusString())
                    add(rainfall.toString())
                    add(baseBiomeRegName ?: "")
                    add(fillerBlock.toString())
                    add(topBlock.toString())
                    client { add(waterColor.toHexColorString()) }
                    add(waterColorMultiplier.toString())
                    add(spawningChance.toString())
                    for (type in EnumCreatureType.values()) {
                        add(getSpawnableList(type).joinToString(separator = System.lineSeparator()))
                    }
                }
                yield(this)
            }
        }
    }
}